ORK1Kit/ORK1Kit/ActiveTasks/ORK1ActiveStepViewController.m

```diff
@@ -381,12 +389,39 @@ - (void)finish {
     }
     if (!self.activeStep.startsFinished) {
         if (self.activeStep.shouldContinueOnFinish) {
-            [self goForward];
+            if (self.activeStep.recordersMustCompleteBeforeAdvancingStep) {
+                [self goForwardOnceRecordersHaveCompleted:NO];
+            } else {
+                [self goForward];
+            }
         }
     }
     [self stepDidFinish];
 }
 
+- (void)goForwardOnceRecordersHaveCompleted:(BOOL)stepDidFinishOnly {
+    dispatch_async(dispatch_get_main_queue(), ^{
+        if ([_recordersRunning count] == 0) {
+            if (!stepDidFinishOnly) {
+                [self goForward];
+            }
+            [self stepDidFinish];
+        } else if (_recordersCheckLoopCount > 10) {
+            NSLog(@"Recorders are still not complete after 10 wait cycles, data may be lost in final result.");
+            if (!stepDidFinishOnly) {
+                [self goForward];
+            }
+            [self stepDidFinish];
+        } else {
+            _recordersCheckLoopCount += 1;
+            dispatch_after(dispatch_time(DISPATCH_TIME_NOW, (int64_t)(0.1 * NSEC_PER_SEC)), dispatch_get_main_queue(), ^{
+                NSLog(@"Waiting for [%lu] recorders to complete...", [_recordersRunning count]);
```

Comment:
Minor: I think it makes more sense to put this NSLog statement before the dispatch_after rather than inside the dispatch block

Reply:
Good catch!